### DISCUSSION TITLE
Where should I put shippedProposals option ?

### DISCUSSION THREAD

**Discussion by qiulang** [babel/babel#16127] [Q&A]
As https://babeljs.io/docs/babel-preset-env#corejs says

> when using useBuiltIns: "usage" you have two different alternatives:
>     set the shippedProposals option to true. 
>     use corejs: { version: "3.8", proposals: true }.

That looks like an either-or setting,  I first set them like these, but I got the error `Module not found: Error: Can't resolve 'core-js/modules/web.dom.iterable.js' in  xxxx`

       "useBuiltIns": "usage",
       "shippedProposals": true
        //"corejs": {
        //        "version": 3.33
        //}

I then put `shippedProposals` inside corejs setting, and it works.

          "useBuiltIns": "usage",
          // "shippedProposals": true,
          "corejs": {
                    "version": 3.33,
                     "shippedProposals": true
          } 

But https://babeljs.io/docs/babel-preset-env#shippedProposals apparently suggests it is a top-level option, so I also changed my setting to that and it works too.

          "useBuiltIns": "usage",
          "shippedProposals": true,
          "corejs": {
                    "version": 3.33,
                     //"shippedProposals": true
          } 

I even hit the error said "[BABEL] /Users/langqiu/Projects/call-center/jssip-emicnet/src/ui/index.js: @babel/preset-env: **'shippedproposals' is not a valid top-level option.**
- Did you mean 'shippedProposals'? (While processing: "/Users/langqiu/Projects/call-center/jssip-emicnet/node_modules/@babel/preset-env/lib/index.js")"

But I can not reproduce that now.

So where should I put `shippedproposals` ?

<img width="1324" alt="Screenshot 2023-11-26 at 19 20 13" src="https://github.com/babel/babel/assets/516243/d188fcfe-6f6a-46f6-9a6e-e94a64a5e923">

**Comment by nicolo-ribaudo** [comment]
`corejs: { version: "3.8", proposals: true }` was correct -- could you share the full stack trace, and your lockfile?

**Reply by qiulang** [reply]
@nicolo-ribaudo As a matter of fact, `corejs: { version: "3.8", proposals: true }`, which is what https://babeljs.io/docs/babel-preset-env#corejs says, is NOT correct because of a bug in corejs 3.8. 

I raised [an issue](https://github.com/zloirock/core-js/issues/1309) to corejs so in my case it should be `corejs: { version: "3.8", proposals: false }` or `corejs: { version: "3.33", proposals: true }` (upgrade to latest 3.33)

The corejs author suggested to me that

> I don't recommend to use proposals: true, it's just for experimenting. shippedProposals is much better and safer for most cases. 

When I tried shippedProposals I hit the problem I listed up and hence the question.

**Comment by JLHwung** [comment]
To clarify, `shippedProposals` is an option for `@babel/preset-env`, while top level option refers to those in the top level of the whole Babel config, such as `presets`.

The following config should work, do notice that the it is recommended that the corejs version be a string rather than a float number.
```json
{
 "presets": ["@babel/preset-env", {
    "useBuiltIns": "usage",
    "shippedProposals": true,
    "corejs": "3.33",
  }] 
}
```


> I even hit the error said "[BABEL] /Users/langqiu/Projects/call-center/jssip-emicnet/src/ui/index.js: @babel/preset-env: 'shippedproposals' is not a valid top-level option.
> Did you mean 'shippedProposals'? (While processing: "/Users/langqiu/Projects/call-center/jssip-emicnet/node_modules/@babel/preset-env/lib/index.js")"
> But I can not reproduce that now.

In this case there is indeed a typo in the provided option name: `shippedproposals` should be `shippedProposals`. (we use camelCase rather than camelcase)

**Reply by qiulang** [reply]
@JLHwung thanks for the reply. What about my second question, i.e. put  `shippedProposals` inside `corejs` ?

```
"useBuiltIns": "usage",
 "corejs": {
                "version": 3.33,
                 "shippedProposals": true
} 
```

I didn't see any error for it and I did see the polyfill codes injected. So can it work too?

`shippedproposals` was a silly mistake, sorry for that.

**Comment by qiulang** [comment]
@JLHwung @nicolo-ribaudo A follow-up question (which is probably a separated question) is I also set `"plugins": [["@babel/plugin-transform-runtime", {}]]` so this is my full .bablerc setting.


        "production": {
            "presets": [
                [
                    "@babel/preset-env",
                    {
                        "targets": {
                               "chrome": "84"
                               "ie": "10"
                        },
                        "modules": "commonjs",
                        "debug": true,
                        "useBuiltIns": "usage",
                        "shippedProposals": true,
                        "corejs": "3.33"
                    }
                ]
            ],
            "plugins": [["@babel/plugin-transform-runtime", {}]]
        }

I know in general, @babel/transform-runtime injecting corejs **without** global pollution while @babel/preset-env does, but I have been using this setting for a long time and I didn't find any problem. And I have an extra "benefit" that the bundled file is smaller. 

I am not sure if the smaller bundled size is the result of what https://babeljs.io/docs/babel-plugin-transform-runtime says, "all of the helpers will reference the module @babel/runtime to avoid duplication across your compiled output."

But the document also says "DANGER, When this plugin is enabled, the useBuiltIns option in @babel/preset-env must not be set. ". This is definitely not what I experienced so far. I have not hit the problem with this setting.
